### PR TITLE
Blacklist commands that cause crashes

### PR DESCRIPTION
**Pull request by tylerkaraszewski**
@quinthar 
cc @coleaeason

~HOLD on: https://github.com/Expensify/Server-Expensify/pull/2290~

Fixes: https://github.com/Expensify/Expensify/issues/63840

This change causes a server to broadcast a message upon crashing, so that other servers can blacklist similar commands from running, preventing cascading crashes of the entire cluster.

## Blacklist format - What gets blacklisted?
A set of command properties has been added to `BedrockCommand`. This is named `blacklistableValues`. A plugin may set any string as a `blacklistableValue`. What this means is tha for a command to be blacklisted, it must have the same values for all of the keys in `blacklistableValues`. Let's use an example:

Pretend we have a problematic command called `SetPolicy`. We set no `blacklistableValues` for this command, and it causes a crash. The blacklisted command, then, is jsut:

```
SetPolicy
```

But say we update Auth to add `accountID` to `blacklistableValues`. Say this command causes a crash on accountID `N`.

Now, the blacklisted command is:

```
SetPolicy
accountID: N
```

Whenever Bedrock receives a `SetPolicy` command, it will check whether `accountID` for that command is `N`, and if so, the command is blacklisted, and returns `500 Blacklisted`.

If more than one key is added to `blacklistableValues`, then *all* of those keys must match to blacklist a command. If *no* keys are added, then all commands with the same methodLine are blacklisted.

The blacklist can be reset by sending the control command `ClearCrashBlacklist`.

## How does it work?

Fundamentally, when a server is about to crash, it creates a `BLACKLIST_COMMAND` command, and populates the body of that command with all of the key/value pairs from the command that caused the crash that are included in `blacklistableValues` for that command. It broadcasts this command to all of its peers.

When a server receives a `BLACKLIST_COMMAND` it bypasses all command queueing and makes an entry in its blacklist for that command and its `blacklistableValues`.

## But really, how does it do that?

We add extra functionality to the signal handler, and when it receives any of SIGSEGV, SIGABRT, SIGFPE, SIGILL, or SIGBUS it will call a function before stacktracing and dying. By default, this function is a no-op, and does nothing. However, in the `sync` and `worker` threads, whenever they dequeue a command, they create a lambda, which is bound to the current command and server state, and set their own signal handler function to that lambda. The signal handler function is stored in thread local storage, and since the signals we're handling here are delivered to the thread that generated them, the signal handler will be able to run the function corresponding to the current command on the correct thread.

What the signal handling function does differs depending on which thread calls it. On the sync thread, this is pretty trivial - it instructs the syncNode to broadcast a message to all peers. When that returns, it exits, and the signal handler resumes killing the program.

On a worker thread, this is somewhat more complex, as the worker cannot write directly to the sync node. What the function does here is generate a message to send, set a flag, and then wait on a condition variable.

The top of the sync thread main loop checks for this flag, and if it's set, it will send the message generated by the worker thread to its peers, then notify the thread waiting on the condition variable.

The worker, now that it's done waiting on the condition variable, will then unblock and finish crashing, but the message will have been sent.

It's entirely possible that anything in this signal handling code will fail, as most of this is not allowed in a signal handler, but the penalty for failure is low, as we're crashing anyway.

## Tests

New automated tests have been added.

### CODE REVIEW

File: BedrockServer.cpp [hunk 1]
```diff
@@ -142,6 +138,17 @@ void BedrockServer::sync(SData& args,
...
     while (!syncNode.shutdownComplete()) {
+        // Check to see if we've received a message to broadcast immediately to all peers (this probably means we're
+        // crashing).
+        if (server._hasBroadcastMessage.load()) {
+            SALERT("Broadcasting message immediately: " << server._broadcastMessage.serialize());
+            server._syncNode->broadcastImmediate(server._broadcastMessage);
+
+            // Done, mark it sent and let the caller continue crashing.
+            server._hasBroadcastMessage.store(0);
+            server._waitForBroadcast.notify_one();
```
Comment (quinthar) [inline BedrockServer.cpp]: Why do we "continue crashing" rather than just `exit();` immediately?

File: BedrockServer.cpp [hunk 2]
```diff
@@ -354,12 +361,16 @@ void BedrockServer::sync(SData& args,
...
-
             // Now we can pull the next command off the queue and start on it.
             command = syncNodeQueuedCommands.pop();
             SINFO("[performance] Sync thread dequeued command " << command.request.methodLine << ". Sync thread has "
                   << syncNodeQueuedCommands.size() << " queued commands.");
 
+            // Set the function that lets the signal handler know which command caused a problem, in case that happens.
+            SSetSignalHandlerDieFunc([&](){
+                server._syncNode->broadcastImmediate(_generateBlacklistMessage(command));
+            });
```
Comment (quinthar) [inline BedrockServer.cpp]: I'm not following the logic here.  So if the worker thread gets a signal it will immediately send the message, but if the sync thread gets a signal it just sets a flag such that the message will be sent later?

File: BedrockServer.cpp [hunk 3]
```diff
@@ -1484,6 +1542,9 @@ void BedrockServer::_control(BedrockCommand& command) {
         suppressCommandPort("SuppressCommandPort", true, true);
     } else if (SIEquals(command.request.methodLine, "ClearCommandPort")) {
         suppressCommandPort("ClearCommandPort", false, true);
+    } else if (SIEquals(command.request.methodLine, "ClearCrashBlacklist")) {
+        unique_lock<decltype(_blackListCommandMutex)> lock(_blackListCommandMutex);
```
Comment (quinthar) [inline BedrockServer.cpp]: The command class it a `CrashBlacklist` but the variable is only `_blacklistedCommands`.  Can you rename it to be `_crashBlacklistedCommands` to be consistent?

File: BedrockServer.cpp [hunk 4]
```diff
@@ -1548,3 +1609,16 @@ void BedrockServer::_beginShutdown(const string& reason, bool detach) {
 bool BedrockServer::backupOnShutdown() {
     return _backupOnShutdown;
 }
+
+SData BedrockServer::_generateBlacklistMessage(const BedrockCommand& command) {
```
Comment (quinthar) [inline BedrockServer.cpp]: I'm not sure why this is its own function; it seems like you would always want to just call this, send immediately, and then -- actually, not exit, just start flushing the buffers until you have been able to confirm you've sent it, and then exit.

**Comment by tylerkaraszewski** [comment]
@quinthar - updated. I renamed a bunch of stuff, the biggest change being taking "blacklist" out of this code and just calling these "crash commands". There are already two different blacklists in bedrockServer, and I thought adding a new one ended up being more confusing with the overloading of that term.

Anyway, I renamed some things and added more commenting to hopefully clarify. Let me give you a high-level overview of what's happening here in case that's helpful:

Each thread sets a function for it's signal handler to call in the case it receives SIGSEGV or similar.

1. On any worker thread, this function sets a flag (the flag is just a pointer to the command it's processing), and then waits to be notified. Once notified, it returns, which causes the signal handler to proceed normally (i.e., crash).
2. On the sync thread, this function called in the signal handler just broadcasts a message to peers, returns, and crashes (If we happened to catch the signal in the middle of sending to a peer, then this message is likely corrupt).
3. Additionally, the sync thread does one more check in it's main loop: If a worker thread has set its flag, it broadcasts a message to peers, and then notifies the worker thread that was waiting, which finishes and dies (see point 1. above).

We're not particularly aggressive in the worker signal handler, because we don't need to be. We set a flag, and we wait. Whatever's corrupt or broken probably doesn't affect the sync thread, so there's no real reason to interrupt it. As long as our command object is still intact, the sync thread will be able to read properties from it, and notify it's peers of the bad command. Then the worker can unblock and finish crashing. It's not perfectly safe, because it's a signal handler, but it does almost nothing - set a flag and wait until the sync thread has sent it.

**Comment by quinthar** [comment]
Ohh very cool.  That design makes sense, very clever.  I'll re-review
shortly. Thanks!

On Oct 12, 2017 5:21 PM, "Tyler Karaszewski" <notifications@github.com>
wrote:

@quinthar <https://github.com/quinthar>
cc @coleaeason <https://github.com/coleaeason>

Review with &w=1, most of the changes are adding one level of indentation.

Fixes: Expensify/Expensify#63840
<https://github.com/Expensify/Expensify/issues/63840>

This change causes a server to broadcast a message upon crashing, so that
other servers can blacklist similar commands from running, preventing
cascading crashes of the entire cluster.
Blacklist format - What gets blacklisted?

A set of command properties has been added to BedrockCommand. This is named
blacklistableValues. A plugin may set any string as a blacklistableValue.
What this means is tha for a command to be blacklisted, it must have the
same values for all of the keys in blacklistableValues. Let's use an
example:

Pretend we have a problematic command called SetPolicy. We set no
blacklistableValues for this command, and it causes a crash. The
blacklisted command, then, is jsut:

SetPolicy

But say we update Auth to add accountID to blacklistableValues. Say this
command causes a crash on accountID N.

Now, the blacklisted command is:

SetPolicy
accountID: N

Whenever Bedrock receives a SetPolicy command, it will check whether
accountID for that command is N, and if so, the command is blacklisted, and
returns 500 Blacklisted.

If more than one key is added to blacklistableValues, then *all* of those
keys must match to blacklist a command. If *no* keys are added, then all
commands with the same methodLine are blacklisted.

The blacklist can be reset by sending the control command
ClearCrashBlacklist.
How does it work?

Fundamentally, when a server is about to crash, it creates a
BLACKLIST_COMMAND command, and populates the body of that command with all
of the key/value pairs from the command that caused the crash that are
included in blacklistableValues for that command. It broadcasts this
command to all of its peers.

When a server receives a BLACKLIST_COMMAND it bypasses all command queueing
and makes an entry in its blacklist for that command and its
blacklistableValues.
But really, how does it do that?

We add extra functionality to the signal handler, and when it receives any
of SIGSEGV, SIGABRT, SIGFPE, SIGILL, or SIGBUS it will call a function
before stacktracing and dying. By default, this function is a no-op, and
does nothing. However, in the sync and worker threads, whenever they
dequeue a command, they create a lambda, which is bound to the current
command and server state, and set their own signal handler function to that
lambda. The signal handler function is stored in thread local storage, and
since the signals we're handling here are delivered to the thread that
generated them, the signal handler will be able to run the function
corresponding to the current command on the correct thread.

What the signal handling function does differs depending on which thread
calls it. On the sync thread, this is pretty trivial - it instructs the
syncNode to broadcast a message to all peers. When that returns, it exits,
and the signal handler resumes killing the program.

On a worker thread, this is somewhat more complex, as the worker cannot
write directly to the sync node. What the function does here is generate a
message to send, set a flag, and then wait on a condition variable.

The top of the sync thread main loop checks for this flag, and if it's set,
it will send the message generated by the worker thread to its peers, then
notify the thread waiting on the condition variable.

The worker, now that it's done waiting on the condition variable, will then
unblock and finish crashing, but the message will have been sent.

It's entirely possible that anything in this signal handling code will
fail, as most of this is not allowed in a signal handler, but the penalty
for failure is low, as we're crashing anyway.
Tests

New automated tests have been added.
------------------------------
You can view, comment on, or merge this pull request online at:

  https://github.com/Expensify/Bedrock/pull/306
Commit Summary

   - First working version
   - Done forget new test file.
   - Remove couts from test
   - Cleanup
   - Black magic - blacklist commands on segv
   - remove debug print statements
   - Add command to clear blacklist
   - Don't die on exceptions

File Changes

   - *M* BedrockCommand.h
   <https://github.com/Expensify/Bedrock/pull/306/files#diff-0> (4)
   - *M* BedrockCommandQueue.h
   <https://github.com/Expensify/Bedrock/pull/306/files#diff-1> (14)
   - *M* BedrockServer.cpp
   <https://github.com/Expensify/Bedrock/pull/306/files#diff-2> (815)
   - *M* BedrockServer.h
   <https://github.com/Expensify/Bedrock/pull/306/files#diff-3> (40)
   - *M* libstuff/SSignal.cpp
   <https://github.com/Expensify/Bedrock/pull/306/files#diff-4> (13)
   - *M* libstuff/libstuff.cpp
   <https://github.com/Expensify/Bedrock/pull/306/files#diff-5> (5)
   - *M* libstuff/libstuff.h
   <https://github.com/Expensify/Bedrock/pull/306/files#diff-6> (2)
   - *M* sqlitecluster/SQLiteNode.cpp
   <https://github.com/Expensify/Bedrock/pull/306/files#diff-7> (9)
   - *M* sqlitecluster/SQLiteNode.h
   <https://github.com/Expensify/Bedrock/pull/306/files#diff-8> (3)
   - *M* test/clustertest/testplugin/TestPlugin.cpp
   <https://github.com/Expensify/Bedrock/pull/306/files#diff-9> (14)
   - *A* test/clustertest/tests/j_BadCommandTest.cpp
   <https://github.com/Expensify/Bedrock/pull/306/files#diff-10> (106)
   - *M* test/lib/BedrockTester.cpp
   <https://github.com/Expensify/Bedrock/pull/306/files#diff-11> (3)

Patch Links:

   - https://github.com/Expensify/Bedrock/pull/306.patch
   - https://github.com/Expensify/Bedrock/pull/306.diff

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/Expensify/Bedrock/pull/306>, or mute the thread

**Comment by righdforsa** [comment]
@quinthar what is the status of your review on this? Super excited to get it into production.

**Comment by tylerkaraszewski** [comment]
I realized there's a race condition in here. It's pretty minor and probably not worth worrying about. We'll successfully serialize `send` calls on peer sockets, but with the current code, it's possible that we delete a send socket if it happens to get disconnected while we're trying to write to it from a different thread. Since this can only happen in this PR if we're already about to crash, this is pretty minor - we'd need to get disconnected from a peer at the precise moment we crashed for this to happen, so I'm not particularly worried about it. I'll fix it correctly in the followup PR to handle synchronize commands from worker threads.

File: BedrockServer.cpp [hunk 5]
```diff
@@ -9,7 +9,20 @@ set<string>BedrockServer::_blacklistedParallelCommands;
 recursive_mutex BedrockServer::_blacklistedParallelCommandMutex;
 
 void BedrockServer::acceptCommand(SQLiteCommand&& command) {
-    _commandQueue.push(BedrockCommand(move(command)));
+    // If the sync node tells us that a command causes a crash, we immediately save that.
+    if(SIEquals(command.request.methodLine, "CRASH_COMMAND")) {
```
Comment (quinthar) [inline BedrockServer.cpp]: Can you update this code to reduce the scope of the `unique_lock` to not include the syslog call?  I wager it can be pretty slow.  So how about something like:

```
SData request;
request.deserialize(command.request.content);
{
    //  Atomically update the blacklisted crash command
    unique_lock<decltype(_crashCommandListMutex)> lock(_crashCommandListMutex);
    _crashCommands.insert(make_pair(request.methodLine, request.nameValueMap));
}
SALERT("Blacklisting command (now have " << _crashCommands.size() << " blacklisted commands): " << request.serialize());
```

File: BedrockServer.h [hunk 6]
```diff
@@ -254,4 +243,23 @@ class BedrockServer : public SQLiteServer {
...
     Port* _commandPort;
+
+    // The following variables all exist to to handle commands that seem to have caused crashes. This lets us broadcast
+    // a command to all peer nodes with information about the crash-causing command, so they can refuse to process it if
+    // it gets sent again (i.e., if an end-user clicks 'refresh' after crashing the first node). Because these can
+    // originate in worker threads, much of this is synchronization code to make sure the sync thread can send this
+    // message before the worker exits.
+
+    // A shared mutex to control access to the list of crash-inducing commands.
+    shared_timed_mutex _crashCommandListMutex;
```
Comment (quinthar) [inline BedrockServer.h]: Can you rename `_crashCommandListMutex` to be `_crashCommandsMutex`?  It's not actually a "list", and might as well just match the variable name it's controlling.

File: libstuff/libstuff.cpp [hunk 7]
```diff
@@ -663,7 +666,7 @@ int SParseHTTP(const char* buffer, size_t length, string& methodLine, STable& na
                 }
 
                 // If not processing a chunked body, then finish up.
-                else if (!SIEquals(nameValueMap["Transfer-Encoding"], "chunked")) {
+                else if (it == nameValueMap.end() || !SIEquals(it->second, "chunked")) {
```
Comment (quinthar) [inline libstuff/libstuff.cpp]: Why this change?

File: BedrockServer.cpp [hunk 8]
```diff
@@ -355,12 +352,19 @@ void BedrockServer::sync(SData& args,
...
                 continue;
             }
-
             // Now we can pull the next command off the queue and start on it.
             command = syncNodeQueuedCommands.pop();
             SINFO("[performance] Sync thread dequeued command " << command.request.methodLine << ". Sync thread has "
                   << syncNodeQueuedCommands.size() << " queued commands.");
 
+            // Set the function that will be called if this thread's signal handler catches an unrecoverable error,
+            // like a segfault. This version is simpler than the one in `worker`, as no synchronization needs to be
```
Comment (quinthar) [inline BedrockServer.cpp]: Is this comment out of date?  The worker and sync thread handler functions seem identical to me.

**Comment by quinthar** [comment]
FYI, tagging this PR `HIGH RISK` -- let's reserve this for a risky deploy window.

File: libstuff/libstuff.cpp (hunk 7 above)
Comment (tylerkaraszewski) [reply libstuff/libstuff.cpp]: Because this pollutes every request with a useless, empty `Transfer-Encoding` header currently. This just cleans that up.

File: BedrockServer.cpp [hunk 9]
```diff
@@ -1385,6 +1427,12 @@ void BedrockServer::_status(BedrockCommand& command) {
         content["version"]  = _version;
         content["host"]     = _args["-nodeHost"];
 
+        {
+            // Make it known if anything is known to cause crashes.
+            shared_lock<decltype(_crashCommandMutex)> lock(_crashCommandMutex);
+            content["crashCommands"] = _crashCommands.size();
```
Comment (coleaeason) [inline BedrockServer.cpp]: can we make this a list of commands instead of the number of commands?

Comment (tylerkaraszewski) [reply BedrockServer.cpp]: Yes, we could, but it's not very straightforward, because a "crash command" is a combination of a methodLine and a set of name/value pairs, which isn't going to be super obvious to look at.
